numpy/linalg/tests/test_linalg.py

```diff
@@ -845,6 +846,108 @@ def test_empty(self):
         assert_equal(norm(array([], dtype=self.dt)), 0.0)
         assert_equal(norm(atleast_2d(array([], dtype=self.dt))), 0.0)
 
+    def test_vector_return_type(self):
+        a = np.array([1, 0, 1])
+
+        exact_types = [
+            np.dtype(_).type for _ in np.typecodes['AllInteger']
+        ]
+
+        inexact_types = [
+            np.dtype(_).type for _ in np.typecodes['AllFloat']
+        ]
+
+        all_types = exact_types + inexact_types
+
+        for each_inexact_types in all_types:
+            at = a.astype(each_inexact_types)
+
+            n = norm(at, -np.inf)
+            assert_(issubclass(n.dtype.type, np.floating))
+            assert_almost_equal(n, 0.0)
+
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", RuntimeWarning)
+                n = norm(at, -1)
+                assert_(issubclass(n.dtype.type, np.floating))
+                assert_almost_equal(n, 0.0)
+
+            n = norm(at, 0)
+            assert_(issubclass(n.dtype.type, np.floating))
+            assert_almost_equal(n, 2)
+
+            n = norm(at, 1)
+            assert_(issubclass(n.dtype.type, np.floating))
+            assert_almost_equal(n, 2.0)
+
+            n = norm(at, 2)
+            assert_(issubclass(n.dtype.type, np.floating))
+            assert_almost_equal(n, 2.0**(1.0/2.0))
+
+            n = norm(at, 4)
+            assert_(issubclass(n.dtype.type, np.floating))
+            assert_almost_equal(n, 2.0**(1.0/4.0))
+
+            n = norm(at, np.inf)
+            assert_(issubclass(n.dtype.type, np.floating))
+            assert_almost_equal(n, 1.0)
+
+    def test_matrix_return_type(self):
+        a = np.array([[1, 0, 1], [0, 1, 1]])
+
+        exact_types = [
+            np.dtype(_).type for _ in np.typecodes['AllInteger']
+        ]
+
+        # float16, float128, and complex256 types are unsupported by `linalg`.
+        # As these types are undefined on some platforms, they need to be
+        # skipped by checking their type codes.
+        inexact_types = [
+            np.dtype(_).type for _ in np.typecodes['AllFloat']
+                                if _ not in ['e', 'g', 'G']
+        ]
+
+        all_types = exact_types + inexact_types
+
+        for each_inexact_types in all_types:
+            at = a.astype(each_inexact_types)
+
+            n = norm(at, -np.inf)
```

Comment:
I guess I am more of a stickler for `i`, `j`, and sometimes `k` being `int`s . The rest of them sort of fall of the map for me. Since learning Python, I find myself using `l` for `list`s that I can't think of a better name for.